pkg/controllers/provisioning/suite_test.go

```diff
@@ -113,6 +115,77 @@ var _ = Describe("Provisioning", func() {
 		Expect(len(nodes.Items)).To(Equal(1))
 		ExpectScheduled(ctx, env.Client, pod)
 	})
+	It("should provision single pod if no other pod is received within the batch idle duration", func() {
+		ExpectApplied(ctx, env.Client, test.NodePool())
+		pod := test.UnschedulablePod()
+		ExpectApplied(ctx, env.Client, pod)
+		prov.Trigger(pod.UID)
+
+		wg := sync.WaitGroup{}
+		ExpectToWait(fakeClock, &wg)
+		result := ExpectSingletonReconciled(ctx, prov)
+		Expect(result.RequeueAfter).ToNot(BeNil())
+	})
+	It("should not extend the timeout if we receive the same pod within the batch idle duration", func() {
+		ctx = options.ToContext(ctx, test.Options(test.OptionsFields{
+			BatchMaxDuration:  lo.ToPtr(10 * time.Second),
+			BatchIdleDuration: lo.ToPtr(5 * time.Second),
+		}))
+		ExpectApplied(ctx, env.Client, test.NodePool())
+		pod := test.UnschedulablePod()
+		ExpectApplied(ctx, env.Client, pod)
+		prov.Trigger(pod.UID)
+		wg := sync.WaitGroup{}
+		wg.Add(1)
+		Expect(fakeClock.HasWaiters()).To(BeFalse())
+		go func() {
+			defer GinkgoRecover()
+			defer wg.Done()
+			fakeClock.Step(3 * time.Second)
+			Expect(fakeClock.HasWaiters()).To(BeTrue())
+			prov.Trigger(pod.UID)
+			fakeClock.Step(3 * time.Second)
+			Eventually(func() bool { return fakeClock.HasWaiters() }).
+				// Caution: if another go routine takes longer than this timeout to
+				// wait on the clock, we will deadlock until the test suite timeout
+				WithTimeout(10 * time.Second).
+				WithPolling(10 * time.Millisecond).
+				Should(BeFalse())
+		}()
+		ExpectSingletonReconciled(ctx, prov)
+	})
+	It("should extend the timeout if we receive a new pod within the batch idle duration", func() {
+		ctx = options.ToContext(ctx, test.Options(test.OptionsFields{
+			BatchMaxDuration:  lo.ToPtr(10 * time.Second),
+			BatchIdleDuration: lo.ToPtr(5 * time.Second),
+		}))
+		ExpectApplied(ctx, env.Client, test.NodePool())
+		pod := test.UnschedulablePod()
+		ExpectApplied(ctx, env.Client, pod)
+		prov.Trigger(pod.UID)
+
+		pod2 := test.UnschedulablePod()
+		ExpectApplied(ctx, env.Client, pod2)
+		wg := sync.WaitGroup{}
+		wg.Add(1)
+		go func() {
+			defer GinkgoRecover()
+			defer wg.Done()
+			fakeClock.Step(3 * time.Second)
+			Expect(fakeClock.HasWaiters()).To(BeTrue())
+			prov.Trigger(pod2.UID)
+			fakeClock.Step(5 * time.Second)
+			Expect(fakeClock.HasWaiters()).To(BeTrue())
```

Comment:
I think you want a `Consistently` here to ensure that you are waiting long enough that the controller is actually still hanging